### PR TITLE
Image classification Networks Updating

### LINKED ISSUE #978: Add networks(VGG/DPN/GoogleNet/ResNet/AlexNet) and update Readme

### PR DESCRIPTION
**Pull request by BigFishMaster**
resolve #978

### CODE REVIEW

File: fluid/image_classification/README.md [hunk 1]
```diff
@@ -1,38 +1,41 @@
...
-wget http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_img_train.tar
-# get validation set
-wget http://www.image-net.org/challenges/LSVRC/2012/nnoupb/ILSVRC2012_img_val.tar
-# prepare directory
-tar xf ILSVRC2012_img_train.tar
-tar xf ILSVRC2012_img_val.tar
-
-# unzip all classes data using unzip.sh
-sh unzip.sh
+wget paddl_imagenet2012_dataset_url/ImageNet2012_dataset.tar
```
Comment (sfraczek) [inline fluid/image_classification/README.md]: Please provide the url

File: fluid/image_classification/README.md [hunk 2]
```diff
@@ -1,38 +1,41 @@
...
+tar xf ImageNet2012_dataset.tar
 ```
 
-2. Download training and validation label files from [ImageNet2012 url](https://pan.baidu.com/s/1Y6BCo0nmxsm_FsEqmx2hKQ)(password:```wx99```). Untar it into workspace ```ILSVRC2012/```. The files include
+**step-2:** Download training and validation label files
+```
+wget paddl_imagenet2012_label_url/ImageNet2012_label.tar
```
Comment (sfraczek) [inline fluid/image_classification/README.md]: Same here

File: fluid/image_classification/README.md [hunk 3]
```diff
@@ -50,38 +53,106 @@ val/ILSVRC2012_val_00000004.jpeg 809
...
+       --batch_size=32 \
+       --class_dim=1000 \
+       --image_shape=3,224,224 \
+       --with_mem_opt=False \
+       --pretrained_model=${path_to_pretrain_model}
+```
 
+## Supported models and performances
 
+Models are trained by starting with learning rate ```0.1``` and decaying it by ```0.1``` after each ```30``` epoches, if not special introduced. Available top-1/top-5 validation accuracy on ImageNet 2012 is listed in table. Pretrained models can be downloaded by clicking related model names.
```
Comment (sfraczek) [inline fluid/image_classification/README.md]: URLs are all pointing to image_classification directory. Clicking doesn't download them.

File: fluid/image_classification/eval.py [hunk 4]
```diff
@@ -1,83 +1,127 @@
...
+import functools
+from models.learning_rate import cosine_decay
+from utility import add_arguments, print_arguments
+import math
 
 parser = argparse.ArgumentParser(description=__doc__)
 add_arg = functools.partial(add_arguments, argparser=parser)
-# yapf: disable
-add_arg('batch_size',       int,   32,        "Minibatch size.")
-add_arg('use_gpu',          bool,  True,      "Whether to use GPU or not.")
```
Comment (sfraczek) [inline fluid/image_classification/eval.py]: At intel we are optimizing the CPU path performance of PaddlePaddle. Please do not remove the CPU support.

File: fluid/image_classification/eval.py [hunk 5]
```diff
@@ -1,83 +1,127 @@
...
 import reader
+import argparse
+import functools
+from models.learning_rate import cosine_decay
+from utility import add_arguments, print_arguments
+import math
 
 parser = argparse.ArgumentParser(description=__doc__)
 add_arg = functools.partial(add_arguments, argparser=parser)
-# yapf: disable
```
Comment (sfraczek) [inline fluid/image_classification/eval.py]: Removing the block markers `# yapf: disable` and `# yapf: enable` has made this parameters defining section hard to read. I think it was better previously.

File: fluid/image_classification/eval.py [hunk 6]
```diff
@@ -1,83 +1,127 @@
...
-    cost = fluid.layers.cross_entropy(input=out, label=label)
-    acc_top1 = fluid.layers.accuracy(input=out, label=label, k=1)
-    acc_top5 = fluid.layers.accuracy(input=out, label=label, k=5)
-    avg_cost = fluid.layers.mean(x=cost)
 
-    inference_program = fluid.default_main_program().clone(for_test=True)
+    # model definition
+    model = models.__dict__[model_name]()
+
+    if model_name in ["GoogleNet"]:
```
Comment (sfraczek) [inline fluid/image_classification/eval.py]: Why not ` if model_name is "GoogleNet":` instead?

File: fluid/image_classification/eval.py [hunk 7]
```diff
@@ -1,83 +1,127 @@
...
+        out = model.net(input=image, class_dim=class_dim)
+        cost = fluid.layers.cross_entropy(input=out, label=label)
+
+        avg_cost = fluid.layers.mean(x=cost)
+        acc_top1 = fluid.layers.accuracy(input=out, label=label, k=1)
+        acc_top5 = fluid.layers.accuracy(input=out, label=label, k=5)
+
+    test_program = fluid.default_main_program().clone(for_test=True)
 
-    place = fluid.CUDAPlace(0) if args.use_gpu else fluid.CPUPlace()
```
Comment (sfraczek) [inline fluid/image_classification/eval.py]: At intel we are optimizing the CPU path performance of PaddlePaddle. Please do not remove the CPU support.

File: fluid/image_classification/eval.py [hunk 8]
```diff
@@ -1,83 +1,127 @@
...
+        t2 = time.time()
+        period = t2 - t1
+        loss = np.mean(loss)
+        acc1 = np.mean(acc1)
+        acc5 = np.mean(acc5)
+        test_info[0].append(loss * len(data))
+        test_info[1].append(acc1 * len(data))
+        test_info[2].append(acc5 * len(data))
+        cnt += len(data)
+        print("total_number:", cnt)
```
Comment (sfraczek) [inline fluid/image_classification/eval.py]: Someone running this script wouldn't know what is this `total_number` representing. How about a more self-explanatory message?

File: fluid/image_classification/infer.py [hunk 9]
```diff
@@ -1,69 +1,90 @@
...
 import reader
+import argparse
+import functools
+from models.learning_rate import cosine_decay
+from utility import add_arguments, print_arguments
+import math
 
 parser = argparse.ArgumentParser(description=__doc__)
 add_arg = functools.partial(add_arguments, argparser=parser)
-# yapf: disable
```
Comment (sfraczek) [inline fluid/image_classification/infer.py]: same as previously.

File: fluid/image_classification/infer.py [hunk 10]
```diff
@@ -1,69 +1,90 @@
...
+
     image = fluid.layers.data(name='image', shape=image_shape, dtype='float32')
-    out = SE_ResNeXt(input=image, class_dim=class_dim, layers=args.num_layers)
-    out = fluid.layers.softmax(input=out)
 
-    inference_program = fluid.default_main_program().clone(for_test=True)
+    # model definition
+    model = models.__dict__[model_name]()
+
+    if model_name in ["GoogleNet"]:
```
Comment (sfraczek) [inline fluid/image_classification/infer.py]: same as previously

File: fluid/image_classification/infer.py [hunk 11]
```diff
@@ -1,69 +1,90 @@
...
+        out, _, _ = model.net(input=image, class_dim=class_dim)
+    else:
+        out = model.net(input=image, class_dim=class_dim)
+
+    test_program = fluid.default_main_program().clone(for_test=True)
+
+    if with_memory_optimization:
+        fluid.memory_optimize(fluid.default_main_program())
 
-    place = fluid.CUDAPlace(0) if args.use_gpu else fluid.CPUPlace()
```
Comment (sfraczek) [inline fluid/image_classification/infer.py]: same as previously

File: fluid/image_classification/train.py [hunk 12]
```diff
@@ -4,278 +4,144 @@
...
     label = fluid.layers.data(name='label', shape=[1], dtype='int64')
-    if args.model is 'se_resnext':
-        out = SE_ResNeXt(input=image, class_dim=class_dim, layers=layers)
-    elif args.model is 'mobile_net':
-        out = mobile_net(img=image, class_dim=class_dim)
+
+    # model definition
+    model = models.__dict__[model_name]()
+
+    if model_name in ["GoogleNet"]:
```
Comment (sfraczek) [inline fluid/image_classification/train.py]: same as previously

**Review by sfraczek — CHANGES_REQUESTED** [review]
This is not a complete review, I concentrated on some things relevant for me.

File: fluid/image_classification/README.md [hunk 13]
```diff
@@ -50,38 +53,106 @@ val/ILSVRC2012_val_00000004.jpeg 809
 val/ILSVRC2012_val_00000005.jpeg 516
 ...
 ```
-**synset_words.txt**: the semantic label of each class.
 
-## Training a model
+## Training a model with flexible parameters
 
-To start a training task, one can use command line as:
+After data preparation, one can start  the training by:
```
Comment (qingqing01) [inline fluid/image_classification/README.md]: Remove the extra blank between `start` and `the`.

File: fluid/image_classification/eval.py (hunk 5 above)
Comment (qingqing01) [reply fluid/image_classification/eval.py]: Yeah, I agree.

File: fluid/image_classification/models/dpn.py [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+            'bw': bw,
+            'r': r
+        }
+        netArg['init_num_filter'] = init_num_filter
+        netArg['init_filter_size'] = init_filter_size
+        netArg['init_padding'] = init_padding
+
+        return netArg
+
+    def DualPathFactory(self,
```
Comment (qingqing01) [inline fluid/image_classification/models/dpn.py]: The function name should be `dual_path_factory` in Google Python style.  https://google.github.io/styleguide/pyguide.html

Same as other files and other variables.

File: fluid/image_classification/models/dpn.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+            'k_R': k_R,
+            'G': G,
+            'k_sec': k_sec,
+            'inc_sec': inc_sec,
+            'bw': bw,
+            'r': r
+        }
+        netArg['init_num_filter'] = init_num_filter
+        netArg['init_filter_size'] = init_filter_size
+        netArg['init_padding'] = init_padding
```
Comment (qingqing01) [inline fluid/image_classification/models/dpn.py]: https://google.github.io/styleguide/pyguide.html

netArg -> net_arg. Please follow the Google Python style.

**Review by qingqing01 — COMMENTED** [review]
Greate work.

File: fluid/image_classification/README.md (hunk 1 above)
Comment (BigFishMaster) [reply fluid/image_classification/README.md]: I replace the URL by the official URL and I use download_imagenet2012.sh to get data.

File: fluid/image_classification/README.md (hunk 2 above)
Comment (BigFishMaster) [reply fluid/image_classification/README.md]: I update the URL in download_imagenet2012.sh

File: fluid/image_classification/README.md (hunk 3 above)
Comment (BigFishMaster) [reply fluid/image_classification/README.md]: I cancel the urls without models, and add right urls to networks with trained models

File: fluid/image_classification/eval.py (hunk 4 above)
Comment (BigFishMaster) [reply fluid/image_classification/eval.py]: I add yapf 'disable' and 'enable'

File: fluid/image_classification/eval.py (hunk 5 above)
Comment (BigFishMaster) [reply fluid/image_classification/eval.py]: I add yapf 'disable' and 'enable'

File: fluid/image_classification/eval.py (hunk 6 above)
Comment (BigFishMaster) [reply fluid/image_classification/eval.py]: I update this as you said

File: fluid/image_classification/eval.py (hunk 7 above)
Comment (BigFishMaster) [reply fluid/image_classification/eval.py]: I add cpu mode in eval.py

File: fluid/image_classification/eval.py (hunk 8 above)
Comment (BigFishMaster) [reply fluid/image_classification/eval.py]: I delete this line

File: fluid/image_classification/infer.py (hunk 9 above)
Comment (BigFishMaster) [reply fluid/image_classification/infer.py]: I fix this.

File: fluid/image_classification/infer.py (hunk 10 above)
Comment (BigFishMaster) [reply fluid/image_classification/infer.py]: I update this as you said

File: fluid/image_classification/infer.py (hunk 11 above)
Comment (BigFishMaster) [reply fluid/image_classification/infer.py]: I fix this.

File: fluid/image_classification/train.py (hunk 12 above)
Comment (BigFishMaster) [reply fluid/image_classification/train.py]: I update this as you said

File: fluid/image_classification/README.md (hunk 13 above)
Comment (BigFishMaster) [reply fluid/image_classification/README.md]: I delete the space

File: fluid/image_classification/models/dpn.py (hunk 14 above)
Comment (BigFishMaster) [reply fluid/image_classification/models/dpn.py]: I change `DualPathFactory` to `dual_path_factory` and revise others which have same problem

**Comment by BigFishMaster** [comment]
@sfraczek Hi,  I have fix the problems you have metioned. Please review this PR.

File: fluid/image_classification/README.md (hunk 1 above)
Comment (sfraczek) [reply fluid/image_classification/README.md]: ok

File: fluid/image_classification/README.md (hunk 2 above)
Comment (sfraczek) [reply fluid/image_classification/README.md]: ok

File: fluid/image_classification/README.md (hunk 3 above)
Comment (sfraczek) [reply fluid/image_classification/README.md]: ok

File: fluid/image_classification/eval.py (hunk 4 above)
Comment (sfraczek) [reply fluid/image_classification/eval.py]: ok

File: fluid/image_classification/eval.py (hunk 5 above)
Comment (sfraczek) [reply fluid/image_classification/eval.py]: ok

File: fluid/image_classification/eval.py (hunk 6 above)
Comment (sfraczek) [reply fluid/image_classification/eval.py]: ok

File: fluid/image_classification/eval.py (hunk 7 above)
Comment (sfraczek) [reply fluid/image_classification/eval.py]: ok

File: fluid/image_classification/eval.py (hunk 8 above)
Comment (sfraczek) [reply fluid/image_classification/eval.py]: ok

File: fluid/image_classification/infer.py (hunk 9 above)
Comment (sfraczek) [reply fluid/image_classification/infer.py]: ok. can you add yapf: disable/enable here too?

File: fluid/image_classification/train.py (hunk 12 above)
Comment (sfraczek) [reply fluid/image_classification/train.py]: ok

File: fluid/image_classification/infer.py (hunk 11 above)
Comment (sfraczek) [reply fluid/image_classification/infer.py]: thanks

File: fluid/image_classification/infer.py (hunk 10 above)
Comment (sfraczek) [reply fluid/image_classification/infer.py]: ok

File: fluid/image_classification/infer.py (hunk 9 above)
Comment (sfraczek) [reply fluid/image_classification/infer.py]: ok

File: fluid/image_classification/eval.py (hunk 8 above)
Comment (sfraczek) [reply fluid/image_classification/eval.py]: ok

**Review by sfraczek — APPROVED** [review]

File: fluid/image_classification/README.md [hunk 16]
```diff
@@ -50,38 +47,111 @@ val/ILSVRC2012_val_00000004.jpeg 809
...
+* color jitter
+* random crop
+* center crop
+* resize
+* flipping
+
+## Finetuning
+
+Finetuning is to finetune model weights in a specific task by loading pretrained weights. After initializing ```path_to_pretrain_model``` , one can finetune a model as:
 ```
```
Comment (qingqing01) [inline fluid/image_classification/README.md]: Need to give the training results here.

File: fluid/image_classification/README_cn.md [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+- [数据准备](#data-preparation)
+- [模型训练](#training-a-model)
+- [参数微调](#finetuning)
+- [模型评估](#evaluation)
+- [模型推断](#inference)
+- [已有模型及其性能](#supported-models)
+
+## 安装
+
+在当前目录下运行样例代码需要PadddlePaddle的v0.10.0或以上的版本。如果你的运行环境中的PaddlePaddle低于此版本，请根据[安装文档](http://www.paddlepaddle.org/docs/develop/documentation/zh/build_and_install/pip_install_cn.html)中的说明来更新PaddlePaddle。
```
Comment (qingqing01) [inline fluid/image_classification/README_cn.md]: PadddlePaddle的v0.10.0 -> PadddlePaddle Fluid v0.13.0

File: fluid/image_classification/README.md [hunk 18]
```diff
@@ -1,38 +1,35 @@
...
+- [Evaluation](#evaluation)
+- [Inference](#inference)
+- [Supported models and performances](#supported-models)
 
-# SE-ResNeXt for image classification
+## Installation
 
-This model built with paddle fluid is still under active development and is not
-the final version. We welcome feedbacks.
+Running sample code in this directory requires PaddelPaddle v0.10.0 and later. If the PaddlePaddle on your device is lower than this version, please follow the instructions in [installation document](http://www.paddlepaddle.org/docs/develop/documentation/zh/build_and_install/pip_install_cn.html) and make an update.
```
Comment (qingqing01) [inline fluid/image_classification/README.md]: PaddelPaddle v0.10.0 -> PaddelPaddle Fluid v0.13.0

File: fluid/image_classification/README_cn.md [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+下面给出了ImageNet分类任务的样例，首先，通过如下的方式进行数据的准备：
+```
+cd data/ILSVRC2012/
+sh download_imagenet2012.sh
+```
+在```download_imagenet2012.sh```脚本中，通过下面两步来准备数据：
+
+**步骤一：** 从ImageNet官网下载ImageNet-2012的图像数据。训练以及验证数据集会分别被下载到"train" 和 "val" 目录中。
+
+**步骤二：** 下载训练与验证集合对应的标签文件。下面两个文件分别包含了训练集合与验证集合中图像的标签：
```
Comment (qingqing01) [inline fluid/image_classification/README_cn.md]: 告诉用户需要注册ImageNet官网，需要获取key？  提醒下用户下载ImageNet时间较长， 如果用户已经有了数据，按照下面方式组织。

同步改下英文。

File: fluid/image_classification/data/ILSVRC2012/download_imagenet2012.sh [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+  filename=`basename $x .tar`
+  mkdir -p $filename
+  tar -xf $x -C $filename
+  rm -rf $x
+done
+cd -
+
+echo "Download imagenet validation data..."
+mkdir -p ${valid_folder}
+#wget -nd -c ${root_url}/${valid_tar}
```
Comment (qingqing01) [inline fluid/image_classification/data/ILSVRC2012/download_imagenet2012.sh]: Remove #

File: fluid/image_classification/data/ILSVRC2012/download_imagenet2012.sh [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+tar xf ${valid_tar} -C ${valid_folder}
+
+echo "Download imagenet label file: val_list.txt & train_list.txt"
+label_file=ImageNet_label.tgz
+label_url=http://imagenet-data.bj.bcebos.com/${label_file}
+#wget -nd -c ${label_url}
```
Comment (qingqing01) [inline fluid/image_classification/data/ILSVRC2012/download_imagenet2012.sh]: Remove #

File: fluid/image_classification/README_cn.md [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+模型推断可以获取一个模型的预测分数或者图像的特征：
+```
+python infer.py \
+       --model=SE_ResNeXt50_32x4d \
+       --batch_size=32 \
+       --class_dim=1000 \
+       --image_shape=3,224,224 \
+       --with_mem_opt=True \
+       --pretrained_model=${path_to_pretrain_model}
+```
```
Comment (qingqing01) [inline fluid/image_classification/README_cn.md:166]: 给个示例，预测结果长啥样。

File: fluid/image_classification/README_cn.md [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+模型评估是指对训练完毕的模型评估各类性能指标。运行如下的命令，可以获得一个模型top-1/top-5精度:
+```
+python eval.py \
+       --model=SE_ResNeXt50_32x4d \
+       --batch_size=32 \
+       --class_dim=1000 \
+       --image_shape=3,224,224 \
+       --with_mem_opt=True \
+       --pretrained_model=${path_to_pretrain_model}
+```
```
Comment (qingqing01) [inline fluid/image_classification/README_cn.md:139]: 给个示例，下载一个已经训练好的模型，这里eval，输出结果长啥样？

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**